src/bucket/Bucket.cpp

```diff
@@ -181,22 +181,79 @@ Bucket::getBucketEntry(LedgerKey const& k)
     return std::nullopt;
 }
 
+void
+Bucket::loadKeys(std::set<LedgerKey, LedgerEntryIdCmp>& keys,
+                 std::vector<LedgerEntry>& result)
+{
+    std::unordered_map<LedgerKey, std::unordered_set<Hash>> lkToTx;
+    std::unordered_map<Hash, uint32_t> txReadBytes;
+    loadKeys(keys, result, lkToTx, txReadBytes);
+}
+
 // When searching for an entry, BucketList calls this function on every bucket.
 // Since the input is sorted, we do a binary search for the first key in keys.
 // If we find the entry, we remove the found key from keys so that later buckets
 // do not load shadowed entries. If we don't find the entry, we do not remove it
 // from keys so that it will be searched for again at a lower level.
 void
-Bucket::loadKeys(std::set<LedgerKey, LedgerEntryIdCmp>& keys,
-                 std::vector<LedgerEntry>& result)
+Bucket::loadKeys(
+    std::set<LedgerKey, LedgerEntryIdCmp>& keys,
+    std::vector<LedgerEntry>& result,
+    std::unordered_map<LedgerKey, std::unordered_set<Hash>>& lkToTx,
+    std::unordered_map<Hash, uint32_t>& txReadBytes)
 {
     ZoneScoped;
+    auto hasTxnWithReadQuota = [&](auto key) {
+        if (lkToTx.empty() || txReadBytes.empty() ||
+            /* classic */ lkToTx.count(key) == 0)
+        {
+            return true;
+        }
+        // Ensure at least one transaction has read quota for this key.
+        for (auto txn : lkToTx[key])
+        {
+            // TODO maybe if (txReadBytes[txn] >=  getIndex().getPageSize())
+            // Ensure txReadBytes[txn] is non-zero
+            if (txReadBytes[txn] > 0)
```

Comment:
I'm unsure what the limit should be here; (txReadBytes[txn] >= getIndex().getPageSize()) is always true for indexes with 0 page size (range index).
Currently this implementation will possibly load unaccounted for extra bytes, as it just ensures the remaining quota is non-zero.

Reply:
`txReadBytes[txn] > 0` is the correct behavior if the `Bucket` index is a range-based index. Suppose a TX wants to read a 4 byte entry, and the page that contains the 4 byte entry is 4 KB. We should still be able to load that entry. In the worst case, we read up to 1 page too much of data per TX, which is fine.

Reply:
>  txReadBytes[txn] > 0 is the correct behavior if the Bucket index is a range-based index.

What is the correct behavior for non range-based indexes? The same?